### DISCUSSION TITLE
Empty mock response when using IO.copy_stream

### DISCUSSION THREAD

**Discussion by avlazarov** [sinatra/sinatra#1871] [Q&A]
I'm using the helper `Sinatra::Streaming` from the `sinatra-contrib` gem, where the yielded object of the `stream` method is meant to be IO-like. Hence I'm using `IO.copy_stream` to feed this IO-like object some data. All works fine when serving via Puma, however when testing the code with [rack-test](https://github.com/rack/rack-test/) the response body is empty. I managed to pull the important parts from `rack-test` and got the minimal amount of code to reproduce the issue:

```
# app.rb
require 'sinatra'
require 'sinatra/streaming'

class App < Sinatra::Base
  helpers Sinatra::Streaming

  get '/copy_stream' do
    stream do |io|
      IO.copy_stream StringIO.new('ok'), io
    end
  end

  get '/write' do
    stream do |io|
      io.write 'ok'
    end
  end
end

# test.rb
require_relative 'app'

def body_for(path)
  env_hash = {
    'rack.test' => true,
    'REQUEST_METHOD' => 'GET',
  }
  env = Rack::MockRequest.env_for "http://localhost:3000/#{path}", env_hash

  status, headers, body = App.call(env).to_a
  response = Rack::MockResponse.new status, headers, body

  response.body
end

body_for('/copy_stream') => ""
body_for('/write') => "ok"
```

As you can see, just doing `io.write 'ok'` works fine, however `IO.copy_stream` for some reason results into an empty mock response body. Does not really matter if the first argument is `StringIO`, `File`, or whatever.

Any ideas?

Sinatra and contrib versions: 3.0.2
Rack version: 2.2.4 
Ruby version: 3.0.3

**Comment by dentarg** [comment]
Converted this to a discussion because it isn't clear to me if something needs to be changed in Sinatra. I don't have any explanation, I think you need to dive deeper yourself. Have you read the rack-test source code?

**Comment by avlazarov** [comment]
@dentarg I'm not using `rack-test` in the example above, I've just mentioned it since it was where I first found the issue. Both `Rack::MockRequest` and `Rack::MockResponse` come from `rack` itself as the way to test without going through the whole stack. I'm expecting Sinatra's `stream` option to work with those two, but that does not seems to be the case when using `IO.copy_stream`.

I can look intro this in details when I have more time, but the fact that `write` works and `IO.copy_steam` (which invokes `#responds_to?(:to_path)` and `#write` on the source IO-like object) not, is odd.
